clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -8007,10 +8007,29 @@ static bool checkSYCLAddIRAttributesMergeability(const AddIRAttrT &NewAttr,
 
 void Sema::CheckSYCLAddIRAttributesFunctionAttrConflicts(Decl *D) {
   const auto *AddIRFuncAttr = D->getAttr<SYCLAddIRAttributesFunctionAttr>();
-  if (!AddIRFuncAttr || AddIRFuncAttr->args_size() == 0 ||
+
+  // If there is no such attribute there is nothing to check. If there are
+  // dependent arguments we cannot know the actual number of arguments so we
+  // defer the check.
+  if (!AddIRFuncAttr ||
       hasDependentExpr(AddIRFuncAttr->args_begin(), AddIRFuncAttr->args_size()))
     return;
 
+  // If there are no name-value pairs in the attribute it will not have an
+  // effect and we can skip the check. The filter is ignored.
+  size_t NumArgsWithoutFilter =
+      AddIRFuncAttr->args_size() - (AddIRFuncAttr->hasFilterList() ? 1 : 0);
+  if (NumArgsWithoutFilter == 0)
+    return;
+
+  // "sycl-single-task" is an implicitly used name which is present on all
+  // single_task invocations. It can only conflict with max_global_work_dim, but
+  // the value will be the same so there is no need for a warning.
+  if (NumArgsWithoutFilter == 2 &&
+      AddIRFuncAttr->getFilteredAttributeNameValuePairs(Context)[0].first ==
+          "sycl-single-task")
```

Comment:
Maybe we need to invent some spec-defined property instead of a magic string for detecting `single_task` invocations?

Reply:
It is currently used for optimizations further down the pipeline. Having it in the spec would mean it would be exposed to the user, which is not what we want for this. An alternative would be to have the frontend identify kernels generated from `single_task` and generate information accordingly, but using these attributes for communicating SYCL header information down is not uncommon.

For this check specifically though, if we are lucky and the compile-time properties ever replace the SYCL attributes, this can be removed entirely as the conflicts won't be possible anymore.